dyn_em/module_ieva_em.F

```diff
@@ -0,0 +1,1276 @@
+                                       ims, ime, jms, jme, kms, kme, &
+                                       its, ite, jts, jte, kts, kte, &
+                                       rk_step
+
+   REAL, DIMENSION( ims:ime, kms:kme , jms:jme ), INTENT(IN   ) ::   u, v, ww, w, ph, phb
+
+   REAL, DIMENSION( ims:ime, kms:kme , jms:jme ), INTENT(INOUT) ::  wwE, wwI
+
+   REAL, DIMENSION( ims:ime, jms:jme ), INTENT(IN   ) :: mut
+
+   REAL, DIMENSION( kms:kme ), INTENT(IN   ) :: rdnw 
+   REAL, DIMENSION( kms:kme ), INTENT(IN   ) :: c1, c2
+
+   REAL, INTENT(IN)    :: dt
+   REAL, INTENT(IN)    :: rdx, rdy
+   REAL , DIMENSION( ims:ime , jms:jme ) ,         INTENT(IN   ) :: msfux, msfuy
+   REAL , DIMENSION( ims:ime , jms:jme ) ,         INTENT(IN   ) :: msfvx, msfvy, msfty
+
+! Local variables
+
+   REAL(KIND=4)    :: cr, cx, cy, c0, c2d, cw_max2, cw_min, cw, cff, rdz
+   INTEGER         :: i, j, k, ii, jj, i_start, i_end, j_start, j_end
+   LOGICAL         :: print_flag = .true.
+   SAVE            :: print_flag
+   
+! Implicit advection parameters based on Shchepetkin (2015)
+! Moved these to the top of the module.
+
+!  REAL, PARAMETER :: alpha_max  = 1.1
+!  REAL, PARAMETER :: alpha_min  = 0.9
+   REAL, PARAMETER :: cmnx_ratio = alpha_min/alpha_max
+   REAL, PARAMETER :: cutoff     = 2.0 - cmnx_ratio
+   REAL, PARAMETER :: r4cmx      = 1.0/(4.0 - 4.0*cmnx_ratio )
+   REAL, PARAMETER :: Ceps       = 0.9
+
+   LOGICAL         :: ieva
+
+! Debug declarations
+
+   REAL,    PARAMETER :: binfrac = 0.1
+   INTEGER, PARAMETER :: binsize = 11
+   INTEGER            :: count(binsize)
+   real(KIND=4)       :: binavg(binsize), crs(binsize)
+
+! Check to see if IEVA is on for this RK3 substep
+
+   ieva = CHK_IEVA(config_flags, rk_step)
+
+! =========================================================================
+! PRINT OUT IEVA INFO
+
+     IF( print_flag ) THEN
+       write(wrf_err_message,*) '----------------------------------------'
+       CALL wrf_debug( 0, wrf_err_message )
+       write(wrf_err_message,*) 'WW_SPLIT:  ZADVECT_IMPLICT = ', config_flags%zadvect_implicit
+       CALL wrf_debug( 0, wrf_err_message )
+       write(wrf_err_message,*) 'WW_SPLIT:  IEVA = ', ieva
+       CALL wrf_debug( 0, wrf_err_message )
+       IF( config_flags%zadvect_implicit > 1 ) THEN
+         write(wrf_err_message,*) 'WW_SPLIT:  WARNING: IEVA IS ONLY RUNNING on LAST RK-substep'
+         CALL wrf_debug( 0, wrf_err_message )
+       ENDIF
+       write(wrf_err_message,*) 'WW_SPLIT:  dt = ', dt
+       CALL wrf_debug( 0, wrf_err_message )
+       write(wrf_err_message,*) 'WW_SPLIT:  alpha_max/min = ', alpha_max, alpha_min
+       CALL wrf_debug( 0, wrf_err_message )
+       print_flag = .false.     
+      ENDIF
+
+! =========================================================================
+   
+   i_start = its-1
+   i_end   = MIN(ite,ide-1) + 1
+   j_start = jts - 1
+   j_end   = MIN(jte,jde-1) + 1
+   
+   wwE(:,:,:) = 0.0
+   wwI(:,:,:) = 0.0
```

Comment:
Lou,
This type of initialization could break parallel bit-for-bit results. It would be better to loop from the tile dims (its, ite, jts, jte, etc). You are initializing halo regions with this array syntax. This is not a problem for local arrays that are already just tile dimensioned - you can initialize the entire tile.